pkg/alicloud/client/client_dns.go

```diff
@@ -35,31 +40,36 @@ func (f *clientFactory) NewDNSClient(region, accessKeyID, accessKeySecret string
 	}
 
 	return &dnsClient{
-		*client,
+		Client:       *client,
+		accessKeyID:  accessKeyID,
+		domainsCache: f.domainsCache,
 	}, nil
 }
 
-// GetDomainNames returns a list of all domain names.
-func (d *dnsClient) GetDomainNames(ctx context.Context) ([]string, error) {
-	var domains []string
-	pageSize, pageNumber := 20, 1
-	req := alidns.CreateDescribeDomainsRequest()
-	req.PageSize = requests.NewInteger(pageSize)
-	for {
-		req.PageNumber = requests.NewInteger(pageNumber)
-		resp, err := d.Client.DescribeDomains(req)
-		if err != nil {
-			return nil, err
-		}
-		for _, domain := range resp.Domains.Domain {
-			domains = append(domains, domain.DomainName)
-		}
-		if resp.PageNumber*int64(pageSize) >= resp.TotalCount {
-			break
-		}
-		pageNumber++
+// GetDomainNames returns a map of all domain names mapped to their composite domain names.
+func (d *dnsClient) GetDomainNames(ctx context.Context) (map[string]string, error) {
+	domains, err := d.getDomainsWithCache()
+	if err != nil {
+		return nil, err
 	}
-	return domains, nil
+	domainNames := make(map[string]string)
+	for _, domain := range domains {
+		domainNames[domain.DomainName] = CompositeDomainName(domain.DomainName, domain.DomainId)
+	}
+	return domainNames, nil
+}
+
+// GetDomainName returns the composite domain name of the domain with the given domain id.
+func (d *dnsClient) GetDomainName(ctx context.Context, domainId string) (string, error) {
+	domains, err := d.getDomainsWithCache()
+	if err != nil {
+		return "", err
+	}
+	domain, ok := domains[domainId]
+	if !ok {
+		return "", fmt.Errorf("DNS domain with id %s not found", domainId)
+	}
+	return CompositeDomainName(domain.DomainName, domain.DomainId), nil
```

Comment:
I would suggest to deprecate `zone` in status, and introduce 2 fields `zoneId` and `zoneName`.  Using `compositeDomainName` is a little tricky.

Reply:
This would be really hard to do. The `DNSRecord` resource is used in g/g and also across 5 different providers, and alicloud is the only one where we would use both `zoneId` and `zoneName` (and only because historically `external-dns-management` uses domain ids that can't be used directly with the Alicloud API). I would like to contain this alicloud-specific issue to alicloud, and avoid changing the API and all other providers just because of it.

Reply:
How about we still only store `domain name` in `zone` field? It seems no side effect. 
It could be simpler and easy to understand in the code.

The purpose of this PR is to handle *union typed* value for `spec.zone`. I think the only place to consider it is in  function `getDomainName`.

Maybe you have concerns about an extra api call to alicloud. But I don't think it is an issue. 
- We already cached it.
- We will later enforce client ratelimit.

You know, we can't avoid AliCloud API call because of the following method `CreateOrUpdateDomainRecords`.

Reply:
> How about we still only store domain name in zone field? It seems no side effect.

The reason I would like to have the domain id as well is to make sure that `DescribeDomains` will be called only if the domain id in `spec.zone` changed (or was added where it was previously missing), or if it's the very first reconciliation and the status is not updated yet. I would not really like to perform a `DescribeDomains` on every call if it can be avoided, no matter that we cache the result. And introducing a client side rate limiter is no real solution either (you will still get errors if the configured rate limit is exceeded, they tend to resolve faster than if you exceed the provider rate limit but they are still very visible and annoying).

> You know, we can't avoid AliCloud API call because of the following method CreateOrUpdateDomainRecords.

This has no impact on this discussion or this PR, it's really only about avoiding redundant calls to `DescribeDomains`.

Reply:
I think if we maintain the domain cache correctly, it doens't have any issue.

Reply:
In my opinion, caching comes with its own set of issues due to the cache potentially becoming outdated (even if for an hour, this can still cause friction). I introduced it there because I don't really know what else to propose to reduce the number of redundant `DescribeDomains` calls, I just hope (but I am not certain) we will not have to revisit it. I still strongly suggest to keep the logic that relies on the `zone` field in status. This is what all other `DNSRecord` controllers do and I see no reason to abandon it here.

What is your concern actually? Having the `zone` field in the format `<domain-name>:<domain-id>` is not ideal, but certainly not a big issue either?

Reply:
It is OK. Just personal preference.

Reply:
OK, let's keep it as it is then, thanks!